### DISCUSSION TITLE
Viability of building modules as single assemblies (rather than Client + Server + Shared) for simple cases

### DISCUSSION THREAD

**Discussion by lanthonyneville** [oqtane/oqtane.framework#2594] [Q&A] [answered]
I have some modules that do not seem to require a "Server" or "Shared" project, only a "Client" project. These are modules that pull all of their data from an external API, so not require their own API hosted inside Oqtane, i.e. in a Server project.

For these modules, I had nothing in the Server project other than the ModuleManager class, and nothing at all in the Shared project.

In order to simplify management and installation, I tried reducing the module to a single project by:
- moving the ModuleManager class into the Client project
- renaming the Client project assembly and namespaces to remove ".Client"

It seems to work - the assembly is recognised by Oqtane, and works on the page (so long as I observe the rules about assembly and namespace names here: https://github.com/oqtane/oqtane.framework/discussions/2570#discussioncomment-4753730)

I am currently running the site in Server mode, but would like the option to try WebAssembly mode in the future. I imaging running my modules in the browser would still work, so long as the external APIs I am using are reachable from the browser.

One concern might be security & IP protection - in WebAssembly mode, there is no longer an option to maintain & run sensitive code on the server, because everything runs in the browser.

**If I want to support both Server and WebAssembly mode**, is there any other reason why this approach might not work (so long as the module is a simple one as above) ?

**If I want to support only Server mode** (which I can because the modules are only for use on a site I control), is there any reason why this approach might not work for all modules?

**Comment by sbwalker** [comment, marked as answer]
I do not see any issues with this approach. If you only have a client-side workload then you do not need the Server project.
